File: 数组/3.二分查找/704.py
'''
Description: 
Author: young
Date: 2022-08-16 10:10:15
LastEditTime: 2022-08-16 10:17:15
FilePath: \young_leetcode\数组\3.二分查找\704.py
'''


# 有序数组 - 二分法
def search(nums,target):
    left = 0
    right = len(nums)-1

    while left <= right:
        mid = int(left+right) / 2
        if (nums[mid] > target):
            right = mid - 1
        elif nums[mid] < target:
            left = mid + 1
        else:
            return mid
    
    return -1
    
    

File: 数组/3.二分查找/704.js
/*
 * @Description: 
 * @Author: young
 * @Date: 2022-08-16 10:17:20
 * @LastEditTime: 2022-08-16 10:23:44
 * @FilePath: \young_leetcode\数组\3.二分查找\704.js
 */

const search = function(nums,target) {
    let left = 0, right = nums.length-1;
    let middle = 0;
    
    while (left <= right) {
        middle = parseInt((left + right) / 2);
        if (nums[middle] > target) {
            right = middle - 1;
        } else if (nums[middle] < target) {
            left = middle + 1;
        } else {
            return middle;
        }
    }

    return -1;
}

// 1.数组无重复元素
// 2.数字已经有序

// 二分法的一个关键在于确定边界条件；此处需要保持区间不变性；要么所有的区间都是左闭右闭，要么所有都是左闭右开
